### PR TITLE
WIP: Typing party

### PR DESCRIPTION
**Pull request by geier**

### CODE REVIEW

**Comment by WhyNotHugo** [comment]
Hint: I've had to add this to `setup.cfg` on pretty much all proyects where I use `mypy`:

```dosini
[mypy]
# Silence warnings given my using untyped libraries:
ignore_missing_imports = True
# See https://github.com/python/mypy/issues/7511:
warn_no_return = False
```

**Comment by geier** [comment]
Protocols are only supported from python 3.8. Is it time to drop support for 3.7 yet?

**Comment by geier** [comment]
Oh, we currently even support 3.6.

**Comment by WhyNotHugo** [comment]
Python 3.6 is past its EOL and can be dropped. I think typing without dropping it might be a lot harder to get right.

**Comment by geier** [comment]
Three issues left, not really sure how to fix them atm.

File: khal/khalendar/vdir.py [hunk 1]
```diff
@@ -29,13 +31,13 @@ def __get__(self, obj, cls):
         return result
 
 
-def to_unicode(x, encoding='ascii'):
+def to_unicode(x: Union[str, bytes], encoding='ascii'):
     if not isinstance(x, str):
         return x.decode(encoding)
     return x
```
Comment (WhyNotHugo) [inline khal/khalendar/vdir.py:37]: Slightly off-topic: this function is a leftover from the python2 days. See https://github.com/pimutils/khal/pull/1155. Sorry for the conflict.

File: khal/khalendar/vdir.py [hunk 2]
```diff
@@ -224,10 +228,9 @@ def upload(self, item):
                 fpath, etag = self._upload_impl(item, href)
             else:
                 raise
-
         return href, etag
 
-    def _upload_impl(self, item, href):
+    def _upload_impl(self, item, href: str):
```
Comment (WhyNotHugo) [inline khal/khalendar/vdir.py:233]: **nitpick**: return type? I believe it's `Tuple[str, str]`.

**Review by WhyNotHugo — COMMENTED** [review]
Those types in `custom_types` are pretty neat!

**Comment by WhyNotHugo** [comment]
Regarding `khal/parse_datetime.py:222: error: Incompatible types in assignment (expression has type "function", variable has type "Callable[..., datetime]")`: I believe this would be fixed by adding typing to the functions assigned to `fun`.

**Comment by WhyNotHugo** [comment]
Nope, that wasn't it.

**Comment by geier** [comment]
I think I got it, some of those functions have defaults for some arguments, others don't. Will fix.

**Comment by WhyNotHugo** [comment]
Aaah, that makes sense!

File: khal/parse_datetime.py [hunk 3]
```diff
@@ -27,36 +27,40 @@
...
 
 
-def datetimefstr(dtime_list, dateformat, default_day=None, infer_year=True,
-                 in_future=True):
+def datetimefstr(
+        dtime_list: List[str],
+        dateformat: str,
+        default_day: Optional[dt.date]=None,
+        infer_year: bool=True,
+        in_future: bool=True,
```
Comment (WhyNotHugo) [inline khal/parse_datetime.py:62]: Eight-space indentation?

**Review by WhyNotHugo — APPROVED** [review]
LGTM.

File: khal/parse_datetime.py (hunk 3 above)
Comment (geier) [reply khal/parse_datetime.py:62]: no idea how that happend...

File: khal/calendar_display.py [hunk 4]
```diff
@@ -22,16 +22,19 @@
...
 
+from .khalendar import CalendarCollection
 from .terminal import colored
 from .utils import get_month_abbr_len
 
 setlocale(LC_ALL, '')
 
 
-def get_weekheader(firstweekday):
+def get_weekheader(firstweekday: int):
```
[the remark this answers is not in the document]
Comment (geier) [reply khal/calendar_display.py]: thanks!

File: khal/khalendar/vdir.py (hunk 2 above)
Comment (geier) [reply khal/khalendar/vdir.py:217]: fixed

**Pull request merged**